### DISCUSSION TITLE
Getting the multiplex setting to work for Revealjs slides

### DISCUSSION THREAD

**Discussion by eteitelbaum** [quarto-dev/quarto-cli#2263] [Q&A (Q1)]
I am not understanding how to get the multiplex setting to work for Revealjs slides. After I have enabled multiplex, e.g. `multiplex: true` and render the presentation, how do I present from the `presentation-speaker.html` file? Am I supposed to just be able to open it in a browser? When I do that, it doesn't look like slides. It just looks like an outline of my talk without any formatting.

**Comment by cderv** [comment]
Documentation for the Multiplex feature from Reveal.js built-in Quarto is here: https://quarto.org/docs/presentations/revealjs/presenting.html#multiplex

>  how do I present from the presentation-speaker.html file

Usually you open the `presentation-speaker.html` on your computer, and your other presentation is online for people to connect. 
When you will change slide on your side, it will change also for viewer looking at `presentation.html`

> Am I supposed to just be able to open it in a browser?

it should work, but you could also serve the file locally. `quarto preview` should do that for you, which is integrated in IDE (RStudio IDE or VS Code)

> When I do that, it doesn't look like slides. It just looks like an outline of my talk without any formatting.

it is probably something with resources not loading. You can check in your browser console (F12) the error you may have. 
If you move the HTML file after rendering, you also need to move the resource folder otherwise, reveal.js and other resources won't be found. Using `self_contained: true` can be an option but can slow down loading: https://quarto.org/docs/presentations/revealjs/presenting.html#publishing

Hope it helps

**Reply by eteitelbaum** [reply]
Thanks. What am I looking for exactly when I hit F12? I see "Inspector", "Console", "Debugger", etc. I don't see anything like an error message but I don't know if I am looking in the right place. 

I do not see a resources folder. Only the `-speaker.html` file. Is there supposed to be a separate folder where the `.qmd` file lives? 

When I render or preview with `multiplex: true`. The speaker view toggle is unresponsive (meaning if I click on it it does nothing).  Perhaps that is a clue as to what is going on?

**Reply by cderv** [reply]
> Thanks. What am I looking for exactly when I hit F12? I see "Inspector", "Console", "Debugger", etc. I don't see anything like an error message but I don't know if I am looking in the right place.

Yes if there was an issue, it should be in Console pane. Resource not found could be seen in a Network pane (maybe not shown by default)

> I do not see a resources folder. Only the -speaker.html file. Is there supposed to be a separate folder where the .qmd file lives?

Did you use `self-contained: true` ? 

> When I render or preview with multiplex: true. The speaker view toggle is unresponsive (meaning if I click on it it does nothing). Perhaps that is a clue as to what is going on?

Yes it seems something is not right. Could you provide a reproducible example of this ? Share a document that don't work for you so I could test ? 

Thanks

**Comment by piiskop** [comment]
The description of using my own multiplex server in _Quarto_ manual is not correct or comprehensive enough. i am describing what i did:

1. i cloned the content from [reveal / multiplex](https://github.com/reveal/multiplex).
2. In the folder where i cloned the data into step 1, i executed `npm install reveal-multiplex`.
3. i executed `node node_modules/reveal-multiplex` and the server started to run at port 1948.
4. i configured my router to direct the traffic to port 1948 into the server machine.
5. i opened the site for [multiplex](http://manual-for-rstudio.tennis24.ee:1948/). It tells: "reveal.js multiplex server.".
6.  i generated tokens and included the following lines in my presentation file:
    ```
    output-file: ../../_book/qmd/presentation.html
    format:
      revealjs:
        incremental: true
        self-contained: true
        multiplex:
          url: 'http://manual-for-rstudio.tennis24.ee/'
          secret: "16759825730498538629"
          socketId: "f3edf0ff0d09982e"
    ```
7. i clicked the button for rendering and it created two files in the output folder: one also for speaker notes.
8. i opened the [presentation](http://manual-for-rstudio.tennis24.ee/presentation.html#/title-slide) in the web browser.
9.  i wanted to open the speaker notes but got the error message instead:
    > presentation.html:1843 Uncaught TypeError: Cannot read properties of undefined (reading 'open')
    > at presentation.html:1843:30
    > at Object.speakerMode (presentation.html:1824:5)
    > at HTMLAnchorElement.onclick (presentation.html:2385:37)

    The speaker notes open if the multiplex information is not there.

Am i doing something wrong? Why does it not work?

**Reply by piiskop** [reply]
i figured out what i was doing wrong and it is not clearly written in the manual: i only did open the main presentation file. However, i have to open the main presentation file on the computer that shows the output to the audience, and the speaker notes file on my smartphone. This way remote controlling works.

**Reply by cderv** [reply]
> However, i have to open the main presentation file on the computer that shows the output to the audience, and the speaker notes file on my smartphone. This way remote controlling works.

We described Multiplex in here: https://quarto.org/docs/presentations/revealjs/presenting.html#multiplex and explain the following 
![image](https://user-images.githubusercontent.com/6791940/218083765-05b1baae-70cc-4fe4-b974-6d4137f4a283.png)

Do you think this should be clarified ? What would be the suggestion that would have help you understand better ?

**Reply by piiskop** [reply]
you are right: the information about these two files is really there. However, i could not find it. Maybe the reason for the latter occasion was that i was stuck in the next section about the server. my suggestion is to enlist the instructions step by step starting with the server setup and configuration as it must happen before it can be used.

As we are discussing *multiplex*, is there a way to use a virtual laser pointer through it?

**Reply by cderv** [reply]
> my suggestion is to enlist the instructions step by step starting with the server setup and configuration as it must happen before it can be used.

Do you mean instruction from https://github.com/reveal/multiplex that should be put in our doc ?

**Reply by piiskop** [reply]
Yes, at least the part about creating your own server.

**Reply by cderv** [reply]
Thanks for the precision. We probably don't include it to avoid having it un-sync with the official doc if things change. We'll look at adding some more information in our doc though. thanks for the feedback

**Comment by turcotte** [comment]
Would it be possible to dynamically obtain the secret, prompting the presenter for it?  If they use a password manager, entering the secret wouldn’t be an issue. This would allow placing `file-speaker.html` in `output_dir`, right?

I really like having `file-speaker.html` on the server. It lets me access my presentation from any device. Here, the restriction would be any device with my password manager installed.

Marcel

**Reply by mcanouil** [reply]
Note that multiplex is not a Quarto feature per se. It's Reveal.s. I doubt Quarto will take on to add dynamic prompting of any kind here as it would switch the burden on Quarto and all the security issues that could arise from this.

Meanwhile, you can set your own server, set secret/id explicitly in your document YAML (from a private repo, etc.)

- https://quarto.org/docs/presentations/revealjs/presenting.html#multiplex-server

**Reply by turcotte** [reply]
Makes sense! Thank you for the reply!
